text/0113-virtual-script-container.md

```diff
@@ -0,0 +1,214 @@
+    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
+    })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
+
+    ga('create', 'UA-XXXXX-Y', 'auto');
+    ga('send', 'pageview');
+</virtual-script>
+```
+
+Inline scripting would also be supported, such as the above example which uses script injection to fetch its required resources.
+
+## Motivation
+
+Web component based applications need the ability to integrate with third party libraries. Because
+web components rely on the Shadow DOM for encapsulation, this development paradigm does not work
+with libraries who are expecting to interact with the application globally. Before there was
+component encapsulation via Shadow DOM it was possible for library authors to:
+
+- Listen for events and know who fired them
+- Define global styling rules that would get applied on every element
+- Query the entire document for any element on the page
+
+These libraries are typically added as global scripts to the document root rather than deeper 
+integrations into the components themselves. When every component in the application has a ShadowRoot attached to them, these libraries will not work as expected.
+
+Some examples of these libraries are:
+
+- **Analytics tools**: Google Tag Manager / Google Analytics, Pendo
+- **Personalization platforms**: Optimizely, Evergage, Google Optimize
+- **Instrumentation tools**: New Relic, Sentry.io
+
+The goal of this proposal is to provide a way for application developers to continue to use their existing integrations until library authors support Shadow DOM integration.
+
+## Design
+
+A new element called `<virtual-script>` encapsulates the various mechanisms to evaluate the scripts in a virtual environment, which solves the problem when using synthetic shadow and subsequently, when using the native Shadow DOM. We can implement this solution in two phases:
+
+### Phase I: Synthetic Shadow
+
+Many LWC applications use the [synthetic-shadow polyfill](https://github.com/salesforce/lwc/tree/master/packages/%40lwc/synthetic-shadow/). This polyfill emulates the native Shadow DOM behavior while still allowing global styles to cascade into the shadow trees. Synthetic Shadow patches many of the DOM APIs that allow you to interact and traverse the DOM tree, (e.g.: `document.querySelector` and `document.querySelectorAll`), these APIs will prevent developers from penetrating the Shadow DOM boundaries defined by their LWC components. In this phase, we want to solve the LWC + Synthetic Shadow scenario.
+
+To ensure we can solve the problem while continuing to guarantee the benefits that are provided by the synthetic-shadow polyfill for all other components running on the main window, we are currently proposing the following characteristics for the virtual-script design:
+
+* Create a JS sandbox before the Synthetic Shadow Polyfill is evaluated.
+  * The sandbox is created using a same-domain iframe.
```

Comment:
Doesn't this present a security concern because of the allowances afforded to same-domain scenarios? While many globals and APIs are restricted, separate domains put up very firm walls that are heavily enforced by the browser.

Reply:
Who would be allowed to use `<virtual-script>` ? It seems like an LWC itself is disallowed from using `<script>`, and it's not generally available, even internally. So this would show up in root HTML documents, which means it's not really for LWC, it's for an app / document bootstrap to define.

Reply:
I see this described in the first paragraph. I suppose there are no additional security concerns unless this is made available to a broader audience, which I suspect security experts would be resistant to.

Reply:
> which I suspect security experts would be resistant to

I'm curious what security vulnerabilities you see beyond what's already possible with a regular <script> tag?

Reply:
Beyond script, nothing. I was imagining a solution that met some "ideal" model of secure, such as running script in a separate domain/iframe, and proxying capabilities through some secure channel, so that it would be safe to let any component use/import anything in the context of its shadow. This solution doesn't try to do that, nor does it need to.

Reply:
yeah, has described above, see @jbenallen question: this is not focus on security, only on the underlaying technology, who can use it in platform is a different question, certainly another RFC.